### DISCUSSION TITLE
Help with VolumePrint

### DISCUSSION THREAD

**Discussion by mooballs** [pschatzmann/arduino-audio-tools#677] [Q&A] [answered]
Hello
I'm trying to access the volume level data from a bluetooth source. The code I have gives me the highest value so far, but doesn't go back down again when the level drops (like when you press the 'hold' function on a multimeter). Please could somebody tell me where I'm going wrong?

#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_sink;
VolumePrint vol_info;

void writeDataStream(const uint8_t *data, uint32_t length) {
  vol_info.write(data, length);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  

  // register A2DP callback
  a2dp_sink.set_stream_reader(writeDataStream, false);

  // Start Bluetooth Audio Receiver
  Serial.print("starting a2dp...");
  a2dp_sink.set_auto_reconnect(true);
  a2dp_sink.start("a2dp");

}

void loop() { 

  Serial.println(vol_info.volume());
 
}

**Comment by pschatzmann** [comment]
I just committed a correction. It seems that this bug has been introduced with the introduction to provide the volume per channel. If you check the volume e.g. on channel 0 with volume(0) it should work with your installed version.

**Comment by mooballs** [comment]
Thanks for replying so quickly Phil.

If I try   Serial.println(vol_info.volume(0));  it just stays at zero, also if I try channel 1

**Reply by pschatzmann** [reply]
Oh, yes - I see, this is because the begin(AudioBaseInfo) method has never been called on the vol_info.

**Reply by mooballs** [reply]
Okay thanks, excuse my ignorance but how exactly do I implement that?
void begin(AudioBaseInfo info) compiles, but I still get 0.00 out, do I need to define any other attributes? 
Or could you point me to an example that uses this function as I can't find it in any?

Thanks again.

**Comment by pschatzmann** [comment, marked as answer]
The critical information that you need to provide is the number of channels. But better try to install the new version and double check if it is working now without begin...

**Reply by mooballs** [reply]
Sorted! Many thanks Phil.
I thought there was an issue when I first tried it as the message 'begin not called' appeared, then I realised I still had the (0) in the vol_info.volume() command.

**Reply by pschatzmann** [reply]
Yes, to get the volume by channel, you would need to tell the class how many channels there are. For getting the overall volume, this is not necessary...

**Comment by mooballs** [comment]
Further to this, when I try to implement VolumePrint in a sketch that also uses VolumeStream (so the overall level can be controlled with a potentiometer) I lose the VolumePrint info. Am I correct in thinking I need to split the stream using MultiOutput? 
Is this somewhere near?

BluetoothA2DPSink a2dp_sink;
A2DPStream in;  
MultiOutput out;
StreamCopy copier(out, in);  //copying in to out
VolumePrint out1;            //changed from vol.info
I2SStream out2;              //changed from i2s
VolumeStream volume(out2);

Then in setup

out.add(out1);
 out.add(out2);

**Comment by pschatzmann** [comment]
No, you just need to put it into the audio chain, if it is before the VolumeStream you get the input volume if it is after, you get the decreased volume ..

**Reply by mooballs** [reply]
Thanks again Phil, the problem is this line:

a2dp_sink.set_stream_reader(read_data_stream, false); 

When I leave it out, I get the vol_info but lose the audio (although I don't understand how turning off the i2s enables the audio). Does this mean I have to wrap the streams up differently?

**Reply by pschatzmann** [reply]
There are many ways but first of all what exectly do you want to achieve ? 
Assuming you want to print the output volume, one options would be
```
                                             -> VolumePrint 
A2DPStream -copy-> VolumeStream -> MultiOutputStream
                                             -> I2SStream
```
So you were basically on the right track and my answer was misleading.

Hoever, you could also just keep the volume and output functionality if A2DP and just implement a callback which prints the volume. This is probably the most efficient way. Look at the basic examples to see how to use the callback. There are 2 different callbacks: one w/o volume applied and one with the volume applied.

**Comment by mooballs** [comment]
I used the stream method you suggested and it works like a dream, thank you.
Basically, using your libraries I'd made a bluetooth player with a volume control that worked beautifully, but wanted to go back and fit a simple VU meter, hence the VolumePrint (didn't need the FFT functionality, otherwise I'd have used that example). 
Many thanks once again.
